cub/block/block_adjacent_difference.cuh

```diff
@@ -200,50 +294,684 @@ public:
+     *   <em>thread</em><sub><em>BLOCK_THREADS</em>-1</sub>).
+     */
+    template <int ITEMS_PER_THREAD,
+              typename OutputT,
+              typename DifferenceOpT>
+    __device__ __forceinline__ void
+    SubtractRight(T (&input)[ITEMS_PER_THREAD],
+                  OutputT (&output)[ITEMS_PER_THREAD],
+                  DifferenceOpT difference_op,
+                  T tile_successor_item)
+    {
+      // Share first item
+      temp_storage.first_items[linear_tid] = input[0];
+
+      CTA_SYNC();
+
+      // Set flag for last thread-item
+      T successor_item = (linear_tid == BLOCK_THREADS - 1)
+                           ? tile_successor_item // Last thread
+                           : temp_storage.first_items[linear_tid + 1];
+
+      #pragma unroll
+      for (int item = 0; item < ITEMS_PER_THREAD - 1; item++)
+      {
+        output[item] = difference_op(input[item], input[item + 1]);
+      }
+
+      output[ITEMS_PER_THREAD - 1] =
+        difference_op(input[ITEMS_PER_THREAD - 1], successor_item);
+    }
+
+    /**
+     * @brief Subtracts the right element of each adjacent pair in range of
+     *        elements partitioned across a CUDA thread block.
+     *
+     * @par
+     * - \rowmajor
+     * - \smemreuse
+     *
+     * @par Snippet
+     * The code snippet below illustrates how to use @p BlockAdjacentDifference to
+     * compute the right difference between adjacent elements.
+     *
+     * @par
+     * @code
+     * #include <cub/cub.cuh>
+     * // or equivalently <cub/block/block_adjacent_difference.cuh>
+     *
+     * struct CustomDifference
+     * {
+     *   template <typename DataType>
+     *   __device__ DataType operator()(DataType &lhs, DataType &rhs)
+     *   {
+     *     return lhs - rhs;
+     *   }
+     * };
+     *
+     * __global__ void ExampleKernel(...)
+     * {
+     *     // Specialize BlockAdjacentDifference for a 1D block of
+     *     // 128 threads on type int
+     *     using BlockAdjacentDifferenceT =
+     *        cub::BlockAdjacentDifference<int, 128>;
+     *
+     *     // Allocate shared memory for BlockDiscontinuity
+     *     __shared__ typename BlockAdjacentDifferenceT::TempStorage temp_storage;
+     *
+     *     // Obtain a segment of consecutive items that are blocked across threads
+     *     int thread_data[4];
+     *     ...
+     *
+     *     // Collectively compute adjacent_difference
+     *     BlockAdjacentDifferenceT(temp_storage).SubtractRightPartialTile(
+     *         thread_data,
+     *         thread_data,
+     *         CustomDifference(),
+     *         valid_items);
+     *
+     * @endcode
+     * @par
+     * Suppose the set of input `thread_data` across the block of threads is
+     * `{ ...3], [4,2,1,1], [1,1,1,1], [2,3,3,3], [3,4,1,4] }`.
+     * and that `valid_items` is `507`. The corresponding output `result` in
+     * those threads will be
+     * `{ ..., [-1,2,1,0], [0,0,0,-1], [-1,0,3,3], [3,4,1,4] }`.
```

Comment:
(continuing from https://github.com/NVIDIA/cub/pull/331#discussion_r674320038)

Gotcha, so the last 5 values are unchanged random values that are ignored by the algorithm.

There's another convention that uses `-` to stand in for such values, e.g. https://github.com/NVIDIA/cub/blob/d6ba3cf0d38e1ab3f079b1f301551e7d4d2ec8e4/cub/device/device_histogram.cuh#L181-L182

I think it's a little easier to parse with the non-numeric characters, since it's immediately clear that they _can't_ be involved in the calculation. If you agree, let's change this to

```
* `{ ...3], [4,2,1,1], [1,1,1,1], [2,3,3,-], [-,-,-,-] }`.
* `{ ..., [-1,2,1,0], [0,0,0,-1], [-1,0,3,-], [-,-,-,-] }`.
```

Reply:
> Gotcha, so the last 5 values are unchanged random values that are ignored by the algorithm.

These values aren't ignored by the algorithm:
> As the documentation states, the input value is copied without modifications if the neighbour value is out of bounds.

If `input` and `output` parameters point to different memory locations, these values will be copied without modification. I'm afraid that non-numeric characters would mean that the output values are unchanged at all, for example:
```
input { 1, 2, 3 };
output { 4, 5, 6 };
valid_items = 0;

adjacent_difference(input, output, valid_items);

// output should be { 1, 2, 3 } and not { 4, 5, 6 }
```

The same with `std::adjacent_difference` [documentation](https://en.cppreference.com/w/cpp/algorithm/adjacent_difference). The output for:
```
std::vector v {2, 4, 6, 8, 10, 12, 14, 16, 18, 20};
```

is 
```
2 2 2 2 2 2 2 2 2 2
```

rather than:

```
- 2 2 2 2 2 2 2 2 2
```

Do you think it'll be clear from the `-` notation that the values are copied?

Reply:
That's a good point -- I'm fine leaving this as-is.